### PR TITLE
Added serializer and deserializer for Random

### LINKED ISSUE #1340: Add serializer and deserializer for Random
in order to save its state

### PR DESCRIPTION
**Pull request by DTanxxx**
Added a RandomConverter file to the serializers folder. I am not entirely confident about the instruction flow for serialization/deserialization when the game is saved/loaded (ie not too sure about when and where are serialization and deserialization actually being called), so any help would be appreciated.  

Closes #1340

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
The JSON converters are registered to be used here: https://github.com/Revolutionary-Games/Thrive/blob/4ba82023cd0169d89ffdc11f7959d60c476ee6de/src/saving/serializers/ThriveJsonConverter.cs#L40

**Comment by hhyyrylainen** [comment]
Which I see you have done already. I would guess that as long as the binary serialization of Random saves the state, this should work already.

**Comment by hhyyrylainen** [comment]
This fails to build because it seems the RandomConverter.cs file is not put in the csproj so it doesn't get compiled.

**Comment by hhyyrylainen** [comment]
I suppose the tar output stream problem is not related to this PR? I'll see if I can get that also to happen on master and try to fix it.

**Comment by hhyyrylainen** [comment]
I can't get the problem to appear locally. I suppose I'll try to fix it half blind.

**Comment by hhyyrylainen** [comment]
Opened a pull request to fix it, I'll merge it once someone can review it (which is hopefully within a day): https://github.com/Revolutionary-Games/Thrive/pull/1817

File: Thrive.csproj [hunk 1]
```diff
@@ -253,14 +254,11 @@
...
-      <Version>12.0.3</Version>
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3">
     </PackageReference>
-    <PackageReference Include="System.Numerics.Vectors">
-      <Version>4.5.0</Version>
+    <PackageReference Include="System.Numerics.Vectors" Version="4.5.0">
     </PackageReference>
-    <PackageReference Include="SharpZipLib">
-      <Version>1.2.0</Version>
+    <PackageReference Include="SharpZipLib" Version="1.3.0">
```
Comment (hhyyrylainen) [inline Thrive.csproj:261]: Ah, I see you updated this package version and that's why CI fails on this PR. I suggest you revert the changes to this section of the csproj file to make CI happy with this PR.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
When testing this, I get this error:
```
Error Saving - Saving failed! An exception happened exception: System.Runtime.Serialization.SerializationException: Type 'System.Random' in Assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' is not marked as serializable.
  at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers (System.RuntimeType type) [0x00045] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.FormatterServices+<>c__DisplayClass9_0.<GetSerializableMembers>b__0 (System.Runtime.Serialization.MemberHolder _) [0x00000] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (TKey key, System.Func`2[T,TResult] valueFactory) [0x00034] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.FormatterServices.GetSerializableMembers (System.Type type, System.Runtime.Serialization.StreamingContext context) [0x0005e] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo () [0x0003d] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00158] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize (System.Object obj, System.Runtime.Serialization.ISurrogateSelector surrogateSelector, System.Runtime.Serialization.StreamingContext context, System.Runtime.Serialization.Formatters.Binary.SerObjectInfoInit serObjectInfoInit, System.Runtime.Serialization.IFormatterConverter converter, System.Runtime.Serialization.Formatters.Binary.ObjectWriter objectWriter, System.Runtime.Serialization.SerializationBinder binder) [0x00006] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize (System.Object graph, System.Runtime.Remoting.Messaging.Header[] inHeaders, System.Runtime.Serialization.Formatters.Binary.__BinaryWriter serWriter, System.Boolean fCheck) [0x0018b] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers, System.Boolean fCheck) [0x0006e] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph, System.Runtime.Remoting.Messaging.Header[] headers) [0x00000] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize (System.IO.Stream serializationStream, System.Object graph) [0x00000] in <46e2faba55964e57bc4a72159b9574b8>:0 
  at RandomConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x0000e] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/RandomConverter.cs:13 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable (Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonConverter converter, System.Object value, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x0008a] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00071] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonObjectContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x0008d] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x000b2] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00079] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerProxy.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type rootType) [0x00008] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonSerializer.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value) [0x00000] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at BaseThriveConverter.WriteMember (System.String name, System.Object memberValue, System.Type memberType, System.Type objectType, Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonSerializer serializer) [0x00017] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/ThriveJsonConverter.cs:583 
  at BaseThriveConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x001c4] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/ThriveJsonConverter.cs:524 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable (Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonConverter converter, System.Object value, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x0008a] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00071] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonObjectContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x0008d] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x000b2] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonObjectContract contract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x0008d] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x000b2] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00079] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonSerializer.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x0023a] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonSerializer.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonConvert.SerializeObjectInternal (System.Object value, System.Type type, Newtonsoft.Json.JsonSerializer jsonSerializer) [0x00028] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonConvert.SerializeObject (System.Object value, System.Type type, Newtonsoft.Json.Formatting formatting, Newtonsoft.Json.JsonSerializerSettings settings) [0x0000e] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at Newtonsoft.Json.JsonConvert.SerializeObject (System.Object value, Newtonsoft.Json.Formatting formatting, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in <2676a2da6edc420e890ed28aa4572ee5>:0 
  at ThriveJsonConverter+<>c__DisplayClass9_0.<SerializeObject>b__0 (Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/ThriveJsonConverter.cs:72 
  at ThriveJsonConverter.PerformWithSettings[T] (System.Func`2[T,TResult] func) [0x00058] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/ThriveJsonConverter.cs:181 
  at ThriveJsonConverter.SerializeObject (System.Object o) [0x0000e] in /home/hhyyrylainen/Projects/thrive/src/saving/serializers/ThriveJsonConverter.cs:72 
  at Save.SaveToFile () [0x00029] in /home/hhyyrylainen/Projects/thrive/src/saving/Save.cs:148 
  at SaveHelper.PerformSave (InProgressSave inProgress, Save save) [0x00002] in /home/hhyyrylainen/Projects/thrive/src/saving/SaveHelper.cs:372 
```

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
So now what this does it:
- override the default Random json serializer (which outputs just `{}`)
- use a binary formatter to make that into a base64 string
- then that base64 string is stored in the json

I don't think this now actually does anything.

**Comment by DTanxxx** [comment]
I think I will need to stick to the BinaryFormatter since json serialization didn't work. Apparently there is no way to serialize System.Random class according to this link: [https://github.com/dotnet/runtime/issues/27429](url)
So I think one solution would be to create our own Random class that incorporates all the System.Random functionalities as well as being marked `Serializable`

**Comment by hhyyrylainen** [comment]
I imagined that it would be possible to use reflection to look into the Random object to get its internal state.
Another approach could be:
At least for newtonsoft json you can externally add an attribute for a class.

**Comment by DTanxxx** [comment]
I tried using reflection but there apparently wasn't any useful internal states that can be used to restore Random. I also looked into adding class attributes with newtonsoft json but it seemed that only json properties can be added (not sure how to add `Serializable`). I found another link on saving the state of Random generator: [https://stackoverflow.com/questions/19512210/how-to-save-the-state-of-a-random-generator-in-c](url)
Following the answer for this post, I added a RandomState struct and saved Random as shown in the example. However I don't think it will work since System.Random is ultimately still not Serializable. Can you please let me know if errors still occur?

**Comment by hhyyrylainen** [comment]
Looking at things in a debugger, I can see the internal state: 
Kuvakaappaus - 2020-11-16 23-20-26

Regarding serialization, did you also specify to look at the private instance fields? I think those fields should be readable and writable like that.

**Comment by hhyyrylainen** [comment]
This answer on github seems to do what I suggest: https://stackoverflow.com/a/19513099/4371508

**Comment by DTanxxx** [comment]
From this article [https://manski.net/2016/08/breaking-nets-random-class/](url) I tried to calculate the current seed using _inext and _inextp fields and serialized that seed, which is later used to instantiate a new Random. The state of Random should be restored if the logic around current seed calculation is correct.

**Comment by hhyyrylainen** [comment]
> To be able to predict the values, one needs to know exactly how Random internally works. This code was run against .NET Framework 4.5.2. If Microsoft ever decides to change the implementation, the prediction won’t work anymore.

Because of that, I'd like to just store the internal fields as is in the json (no need to use binary formatter), to make this future proof against that kind of thing.

File: Thrive.csproj [hunk 2]
```diff
@@ -253,14 +254,11 @@
...
-      <Version>12.0.3</Version>
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3">
     </PackageReference>
-    <PackageReference Include="System.Numerics.Vectors">
-      <Version>4.5.0</Version>
+    <PackageReference Include="System.Numerics.Vectors" Version="4.5.0">
     </PackageReference>
-    <PackageReference Include="SharpZipLib">
-      <Version>1.3.0</Version>
+    <PackageReference Include="SharpZipLib" Version="1.3.0">
```
Comment (hhyyrylainen) [inline Thrive.csproj:261]: Please revert these changes, as this separate `Version` xml node version has seriously messed with package installs on my computer.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

**Comment by hhyyrylainen** [comment]
I went and fixed the deserialization, and added backwards compatibility.
Did you even test it? It took me 15 seconds to go into game, save, and then get the deserialize problem on loading. It's a waste of my time to review PRs that don't even basically do what they should...

Also did some cleanup.

File: Thrive.csproj [hunk 3]
```diff
@@ -261,7 +262,7 @@
       <Version>4.5.0</Version>
     </PackageReference>
     <PackageReference Include="SharpZipLib">
-      <Version>1.3.0</Version>
+      <Version>1.2.0</Version>
```
Comment (hhyyrylainen) [inline Thrive.csproj:265]: I moved over our code to actually use version 1.3.0 of this library

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

**Comment by DTanxxx** [comment]
I didn't test the code. Sorry about that. Currently for my local repo I get this error `Error CS1503: Argument 2: cannot convert from 'System.Text.Encoding' to 'int' (CS1503) (Thrive)` in `Save.cs` from line `using var tar = new TarOutputStream(gzoStream, Encoding.UTF8);` when I build the project so I got lazy and didn't look into the problem and as a result didn't test serialization & deserialization. Just wondering, was the deserialization problem due to this line `var seedArray = item["seedArray"].Value<int[]>();`?

**Comment by hhyyrylainen** [comment]
That's because I already updated master for SharpZipLib 1.3.0. You need to undo that change to the csproj to get the same version as on the master branch.

**Comment by hhyyrylainen** [comment]
>  Just wondering, was the deserialization problem due to this line var seedArray = item["seedArray"].Value<int[]>();

Yes, that always failed with an exception.

**Review by hhyyrylainen — APPROVED** [review]

**Comment by DTanxxx** [comment]
There was another error when I tried to load previous saved files. Was that due to those files being outdated compared to Thrive's version? There were exclamation marks shown for those old files.

**Comment by hhyyrylainen** [comment]
Any save from 0.5.2 and before is utterly incompatible. If the save was made with 0.5.3-alpha (starting from the first commit that changed the version number) it should be compatible.

**Comment by DTanxxx** [comment]
I see, it has been a while since I last played and saved the game so my saves were incompatible. Thanks.

**Comment by hhyyrylainen** [comment]
@DTanxxx  we are in the process of getting a CLA in-place for Thrive in order to make it possible to have Thrive on Steam. Could you maybe sign the CLA? You can do that here: https://dev.revolutionarygamesstudio.com/cla

**Pull request merged**